fleetconfig-controller/internal/controller/v1beta1/spoke_handler.go

```diff
@@ -0,0 +1,1076 @@
+		spoke.SetConditions(true, v1beta1.NewCondition(
+			msg, v1beta1.SpokeJoined, metav1.ConditionFalse, metav1.ConditionTrue,
+		))
+		return errors.New(msg)
+	}
+
+	// spoke cluster has joined successfully
+	spoke.SetConditions(true, v1beta1.NewCondition(
+		"Joined", v1beta1.SpokeJoined, metav1.ConditionTrue, metav1.ConditionTrue,
+	))
+
+	// Label the spoke ManagedCluster if in hub-as-spoke mode.
+	// This allows the 'spoke' ManagedClusterSet to omit the hub-as-spoke cluster from its list
+	// of spoke clusters.
+	if managedCluster != nil && spoke.IsHubAsSpoke() {
+		if managedCluster.Labels == nil {
+			managedCluster.Labels = make(map[string]string)
+		}
+		managedCluster.Labels[v1beta1.LabelManagedClusterType] = v1beta1.ManagedClusterTypeHubAsSpoke
+		if err := common.UpdateManagedCluster(ctx, clusterClient, managedCluster); err != nil {
+			return err
+		}
+		logger.V(0).Info("labeled ManagedCluster as hub-as-spoke", "name", spoke.Name)
+	}
+
+	err = r.deleteKubeconfigSecret(ctx, spoke)
+	if err != nil {
+		logger.Error(err, "warning: failed to remove spoke's kubeconfig secret", "spoke", spoke.Name)
+		return err
+	}
+
+	if !spoke.IsHubAsSpoke() {
+		adc := &addonv1alpha1.AddOnDeploymentConfig{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      v1beta1.FCCAddOnName,
+				Namespace: spoke.Name,
+			},
+			Spec: addonv1alpha1.AddOnDeploymentConfigSpec{
+				AgentInstallNamespace: os.Getenv(v1beta1.ControllerNamespaceEnvVar),
+				CustomizedVariables: []addonv1alpha1.CustomizedVariable{
+					{
+						Name:  v1beta1.HubNamespaceEnvVar,
+						Value: spoke.Spec.HubRef.Namespace,
+					},
+					{
+						Name:  v1beta1.SpokeNamespaceEnvVar,
+						Value: spoke.Namespace,
+					},
+				},
+			},
+		}
+		_, err = addonC.AddonV1alpha1().AddOnDeploymentConfigs(spoke.Name).Create(ctx, adc, metav1.CreateOptions{})
+		if err != nil && !kerrs.IsAlreadyExists(err) {
+			return err
+		}
+
+		err = r.bindAddonAgent(ctx, spoke)
+		if err != nil {
+			return err
+		}
+	}
+
+	enabledAddons, err := handleSpokeAddons(ctx, addonC, spoke)
+	if err != nil {
+		msg := fmt.Sprintf("failed to enable addons for spoke cluster %s: %s", spoke.Name, err.Error())
+		spoke.SetConditions(true, v1beta1.NewCondition(
+			msg, v1beta1.AddonsConfigured, metav1.ConditionFalse, metav1.ConditionTrue,
+		))
+		return err
+	}
+	spoke.Status.EnabledAddons = enabledAddons
+	return nil
+}
+
+func (r *SpokeReconciler) createAgentNamespace(ctx context.Context, spokeName string, spokeKubeconfig []byte) error {
+	logger := log.FromContext(ctx)
+	spokeRestCfg, err := kube.RestConfigFromKubeconfig(spokeKubeconfig)
+	if err != nil {
+		return err
+	}
+	spokeCli, err := client.New(spokeRestCfg, client.Options{})
+	if err != nil {
+		return err
+	}
+	agentNamespace := os.Getenv(v1beta1.ControllerNamespaceEnvVar) // manager.go enforces that this is not ""
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: agentNamespace,
+		},
+	}
+	err = spokeCli.Create(ctx, ns)
```

Comment:
here I would say it's fine to not handle the upsert logic, since all that matters is that the ns exists